src/typ/environment.rs

```diff
@@ -21,24 +22,32 @@ pub struct Environment<'a, 'b> {
     // Accessors defined in the current module
     pub accessors: HashMap<String, AccessorsMap>,
 
-    // Type constructors that have imported or privately defined
-    // but have not yet been used.
-    pub unused_private_types: HashMap<String, SrcSpan>,
-
-    // Constructors that have imported or privately defined but
-    // have not yet been used.
-    // This hashmap contains names shared by a type constructor
-    // AND a value constructor.
-    // i.e. type X { X }
-    pub unused_private_mixed_constructors: HashMap<String, SrcSpan>,
-
     // Warnings
     pub warnings: &'a mut Vec<Warning>,
 
     // Functions that have not yet been inferred then generalised.
     // We use this to determine whether functions that call this one
     // can safely be generalised.
     pub ungeneralised_functions: HashSet<String>,
+
+    // value_usages is a stack of value scopes. When a value is used
+    // it crawls up the scope stack for a value with that name and
+    // increments its usage.
+    pub value_usages: VecDeque<HashMap<String, (ValueKind, SrcSpan, usize)>>,
+}
+
+// For Keeping track of value usages and knowing which error to display.
+#[derive(Debug, Clone)]
+pub enum ValueKind {
```

Comment:
I wonder if there's a better name than `ValueKind` as it is for both types and values.

- `Named`
- `Defined`
- `VariableKind`
- ???

Same goes for the `value_usages`.

Reply:
Hows `EntityKind` and `entity_usages` sound?

Reply:
That's great!